optimism/src/main.rs

```diff
@@ -152,3 +167,71 @@ pub fn main() -> ExitCode {
     // TODO: Logic
     ExitCode::FAILURE
 }
+
+pub fn test_preimage_read(preimage_key_dir: String, configuration: VmConfiguration) -> ExitCode {
+    use rand::Rng;
+
+    let mut po = PreImageOracle::create(&configuration.host);
+    let _child = po.start();
+    println!("Let server start");
+    std::thread::sleep(std::time::Duration::from_secs(5));
+
+    println!("Reading from {}", preimage_key_dir);
+    // Get all files under the preimage db dir
+    let paths = std::fs::read_dir(preimage_key_dir)
+        .unwrap()
+        .map(|res| res.map(|e| e.path()))
+        .collect::<Result<Vec<_>, io::Error>>()
+        .unwrap();
+
+    // Just take 10 elements at random to test
+    let how_many = 10_usize;
+    let max = paths.len();
+    let mut rng = rand::thread_rng();
+
+    let mut selected: Vec<PathBuf> = vec![PathBuf::new(); how_many];
+    for pb in selected.iter_mut().take(how_many) {
+        let idx = rng.gen_range(0..max);
+        *pb = paths[idx].to_path_buf();
+    }
+
+    for (idx, path) in selected.into_iter().enumerate() {
+        let preimage_key = Path::new(&path)
+            .file_name()
+            .unwrap()
+            .to_str()
+            .unwrap()
+            .split('.')
+            .collect::<Vec<_>>()[0];
+
+        let hash = PreimageKey::from_str(preimage_key).unwrap();
+        let key = Key::Keccak(hash.0);
+
+        debug!(
+            "Generating OP Keccak key for {} at index {}",
+            preimage_key, idx
+        );
+
+        let expected = std::fs::read_to_string(path).unwrap();
+
+        debug!("Asking for preimage");
+        let preimage = po.get_preimage(key);
+        let got = hex::encode(preimage.get()).to_string();
+
+        assert_eq!(expected, got);
+    }
+    ExitCode::SUCCESS
+}
+
+pub fn main() -> ExitCode {
```

Comment:
For a future improvement (open follow-up): I would make explicit the split between test and real environment.

Reply:
Indeed. Yeah sure.

Reply:
Opened https://github.com/o1-labs/proof-systems/issues/1486